include/sys/dbuf.h

```diff
@@ -442,10 +445,34 @@ dbuf_find_dirty_eq(dmu_buf_impl_t *db, uint64_t txg)
 	(dbuf_is_metadata(_db) &&					\
 	((_db)->db_objset->os_primary_cache == ZFS_CACHE_METADATA)))
 
-#define	DBUF_IS_L2CACHEABLE(_db)					\
-	((_db)->db_objset->os_secondary_cache == ZFS_CACHE_ALL ||	\
-	(dbuf_is_metadata(_db) &&					\
-	((_db)->db_objset->os_secondary_cache == ZFS_CACHE_METADATA)))
+/*
+ * We want to exclude buffers that are on a special allocation class from
+ * L2ARC.
+ */
+static inline boolean_t dbuf_is_l2cacheable(dmu_buf_impl_t *db)
+{
+	blkptr_t *bp = db->db_blkptr;
+	spa_t *spa = db->db_objset->os_spa;
+	uint64_t vdev = DVA_GET_VDEV(db->db_blkptr->blk_dva);
+	vdev_t *vd = NULL;
+
+	if (bp != NULL) {
+		vdev_t *rvd = spa->spa_root_vdev;
+
+		if (vdev < rvd->vdev_children) {
+			ASSERT(rvd->vdev_child[vdev] != NULL);
+			vd = rvd->vdev_child[vdev];
+		}
+	}
+
+	if ((db->db_objset->os_secondary_cache == ZFS_CACHE_ALL ||
+	    (dbuf_is_metadata(db) &&
+	    db->db_objset->os_secondary_cache == ZFS_CACHE_METADATA)) &&
+	    !(vd != NULL && vd->vdev_alloc_bias == VDEV_BIAS_SPECIAL))
+		return (B_TRUE);
+
+	return (B_FALSE);
+}
```

Comment:
Looking at how this turned out and is called we could even move this whole thing in to `dbuf.c`, say right after `dbuf_is_metadata`.